File: epi/queue.cpp
#include <cstddef>
#include <iostream>
#include <stdexcept>
#include <vector>

// section: interview
// difficulty: easy
// tags: optiver

// constraints
// don't use C++ list/deque's push_back(), etc.

// solution: epi using vector
const size_t MAX_SIZE = 100;

template
<typename T>
class MyQueue {
public:
    MyQueue()
    : d_capacity(MAX_SIZE)
    , d_front(0)
    , d_back(0)
    , d_size(0)
    {
        d_queue.reserve(d_capacity);
    }

    MyQueue(size_t capacity)
    : d_capacity(capacity)
    , d_front(0)
    , d_back(0)
    , d_size(0)
    {
        d_queue.reserve(d_capacity);
    }

    void push(const T& value)
    {
        if (d_size == d_capacity) {
            // shift left
            std::rotate(d_queue.begin(), d_queue.begin() + d_front, d_queue.end());
            d_front = 0;
            d_back = d_size;

            // double size
            d_capacity *= 2;
            d_queue.resize(d_capacity);
        }

        d_queue[d_back++] = value;
        ++d_size;
    }

    T pop()
    {
        if (empty()) {
            throw std::length_error("empty queue");
        }

        int value = d_queue[d_front];
        d_queue[d_front++] = 0;
        --d_size;

        return value;
    }

    T front() const
    {
        if (empty()) {
            throw std::length_error("empty queue");
        }

        return d_queue[d_front];
    }

    bool empty() const { return d_size == 0; }

    size_t size() const { return d_size; }

private:
    std::vector<T> d_queue;
    size_t d_capacity;
    size_t d_front;
    size_t d_back;
    size_t d_size;
};

File: epi/queue.py
#!/usr/bin/env python3

import sys
import unittest

# section: interview
# difficulty: easy
# tags: optiver

# constraints
# don't use Python list/deque's append(), etc.

class MyQueue:
    def __init__(self, capacity=MAX_SIZE) -> None:
        self.__capacity = capacity
        self.__queue = [None] * capacity
        self.__size = self.__front = self.__back = 0

    def __increase_capacity(self, factor=2) -> None:
        self.__capacity *= factor
        self.__queue[:] = self.__queue[:self.__capacity] + \
                [None for _ in range(self.__capacity - len(self.__queue))]

    def push(self, value: int) -> None:
        if self.__size == self.__capacity:
            self.__increase_capacity()

        self.__queue[self.__back] = value
        self.__back += 1
        self.__size += 1

    def pop(self) -> int:
        if self.empty():
            return None

        value = self.__queue[self.__front]
        self.__queue[self.__front] = None
        self.__front += 1
        self.__size -= 1

        return value

    def front(self):
        if self.empty():
            return None

        return self.__queue[self.__front]

    def empty(self):
        return not self.__size

class TestMyQueue(unittest.TestCase):
    def test(self):
        q = MyQueue(1)                  # [None]
        self.assertTrue(q.empty())
        self.assertFalse(q.pop())
        q.push(1)                       # [1]
        q.push(2)                       # [1,2]
        q.push(3)                       # [1,2,3]
        self.assertEqual(q.front(), 1)
        self.assertEqual(q.pop(), 1)    # [2]
        self.assertFalse(q.empty())

if __name__ == '__main__':
    sys.exit(unittest.main())
